### ISSUE TITLE
“\s” in URI regexp is ambiguous

### ISSUE THREAD

**Issue by FGasper** [wamp-proto/wamp-proto#275]
The following Python regexp is given to test URIs for validity:
```
re.compile(r"^([^\s\.#]+\.)*([^\s\.#]+)$")
```
… however, it’s not specified here whether ```\s``` is understood to be with or without the UNICODE flag.

**Comment by FGasper** [comment]
… and, in general, the term “whitespace” doesn’t appear to be qualified anywhere to a precise set of patterns. Different contexts define “whitespace” differently, so to avoid incompatibilities among implementations it seems important to define exactly what is (and is not) meant by this term.

**Comment by oberstet** [comment]
Yeah, ambiguous. I see, you are coming from Perl;) Attention to detail with "text" .. he. This is welcome!

It'll be fun though.

Unicode has a character property "WSpace=Y" defined for exactly 25 characters:

* http://www.unicode.org/Public/UCD/latest/ucd/PropList.txt
* https://en.wikipedia.org/wiki/Unicode_character_property#Whitespace

FWIW, Python disagrees with that:

```console
>>> ws = [hex(cp) for cp in range(65535) if chr(cp).isspace()]
>>> len(ws)
29
>>> ws
['0x9', '0xa', '0xb', '0xc', '0xd', '0x1c', '0x1d', '0x1e', '0x1f', '0x20', '0x85', '0xa0', '0x1680', '0x2000', '0x2001', '0x2002', '0x2003', '0x2004', '0x2005', '0x2006', '0x2007', '0x2008', '0x2009', '0x200a', '0x2028', '0x2029', '0x202f', '0x205f', '0x3000']
```

This is a strict superset of the 25 whitespace characters as defined in Unicode version 9.

The additional 4 ones are: `'0x1c', '0x1d', '0x1e', '0x1f'`.

Eg `001C` is File separator: http://www.fileformat.info/info/unicode/char/001c/index.htm

As can be seen in this latter page, Java also claims `Character.isWhitespace(001C) == true`

**Comment by oberstet** [comment]
Turns out, Python has a very subtle definition of "whitespace" (https://docs.python.org/3/library/stdtypes.html#str.isspace)

Whitespace according to Python is EITHER a character with the `wspace=true` property as from [this](http://www.unicode.org/Public/UCD/latest/ucd/PropList.txt) OR with bidirectional property being one of “WS”, “B”, or “S” according to [this](http://www.unicode.org/Public/UNIDATA/UnicodeData.txt).

I had a long chat on IRC, and this seems to be by design. And it differs from the Perl definition (https://perldoc.perl.org/perlrecharclass.html#Whitespace). Perl defines whitespace _exactly_ as Unicode 9.0: the 25 characters listed there.

I haven't looked deeply into Java, but I expect it to be messed up.

---

Well, this was expected. It's messy.

Right now, I lean towards the Perl definition to be honest. It can be understood by mere mortals;)

Or we define whitespace to be `['0x9', '0xa', '0xb', '0xc', '0xd']` - essentially, everything whitespace in the old days (US-ASCII-7).
